tests/processes/test_wps_package.py

```diff
@@ -230,10 +250,138 @@ def test_stdout_stderr_logging_for_commandline_tool_exception(caplog):
         with open(expect_log, mode="r", encoding="utf-8") as file:
             job_err = file.read()
         log_err = stderr.getvalue() + "\n" + caplog.text + "\n" + job_err
-        assert "Could not retrieve any internal application log." in log_err, (
-            "Since command did not run, nothing captured is expected"
-        )
+        assert "Package completed with errors." in log_err
         assert "Traceback (most recent call last):" in log_err
         assert "weaver.processes.wps_package|mock-process" in log_err
     else:
         pytest.fail("\"wps_package._handler()\" was expected to throw \"PackageExecutionError\" exception")
+
+
+def _combine(dict1, dict2):
+    # type: (Dict[KT, VT_co], Dict[KT, VT_co]) -> Dict[KT, VT_co]
+    dict1 = dict1.copy()
+    dict1.update(dict2)
+    return dict1
+
+
+def assert_equal_requirements_any_order(result, expected):
+    for field in ["hints", "requirements"]:
+        if field in expected:
+            if isinstance(expected[field], dict):
+                assert result[field] == expected[field]
+            else:
+                assert_equal_any_order(result[field], expected[field])
+        else:
+            assert field not in result
+
+
+@pytest.mark.parametrize("original, expected", [
+    (
+        {"requirements": {CWL_REQUIREMENT_APP_DOCKER_GPU: {"dockerPull": "python:3.7-alpine"}}},
+        {"requirements": {CWL_REQUIREMENT_APP_DOCKER: {"dockerPull": "python:3.7-alpine"},
+                          CWL_REQUIREMENT_CUDA: CWL_REQUIREMENT_CUDA_DEFAULT_PARAMETERS}},
+    ),
+    (
+        {"requirements": {CWL_REQUIREMENT_APP_DOCKER_GPU: {"dockerPull": "python:3.7-alpine"},
+                          CWL_REQUIREMENT_CUDA: {"custom": 1, "cudaVersionMin": "11.0", "cudaDeviceCountMin": 8}}},
```

Comment:
@mr-c 
Can you confirm this case? The `cwltool:CUDARequirement` is only permitted in `hints`? 
If so, that's another thing that could be specified in https://github.com/common-workflow-language/cwl-v1.2/issues/212

Reply:
`cwltool:CUDARequirement` should almost always be under `requirements` unless it truly is optional. When using `cwltool` one needs to pass `--enable-ext`. See https://github.com/common-workflow-language/cwltool/blob/0099f11d5a8740d3ca976cd40681523d1002e502/cwltool/main.py#LL661C15-L661C24 
You could choose to permanently enable some of the extensions in weaver; it is up to you if you want a command-line flag or not.

Reply:
Is there an equivalent of `--enable-ext` directly with the `cwltool` code rather than the CLI? 

We use the python factory with loading/runtime context and following parameters:
https://github.com/crim-ca/weaver/blob/master/weaver/processes/wps_package.py#L1288-L1309

Reply:
Currently `setup_schema` (which handles `--enable-ext`) is only called from the CLI loop. So you'll need to call `use_custom_schema()` at startup before using any of the factory methods.

If you don't want to enable all extensions, then make your own copy of `extensions{,-v1.1,-v1.2}.yml`; otherwise you can use the copy inside the `cwltool` module.

If you'd like, we can refactor `setup_schema` to take in a "enable_ext" boolean instead of an entire `argparse.Namespace`

Reply:
Yes, the refactor could be useful to make it accessible. Maybe having some list of enabled extensions or a `.yml` file reference with extensions would be more useful than the boolean. 
I'll look into using `use_custom_schema` in the meantime. Thanks!

Reply:
@mr-c  
I have tried using `use_custom_schema` (https://github.com/crim-ca/weaver/pull/510) as suggested, but I end up having the following error when the CWL document is loaded by the factory.

```
schema_salad.exceptions.ValidationException: http://commonwl.org/cwltool:2:3:  checking object `http://commonwl.org/cwltool#CUDARequirement`
http://commonwl.org/cwltool:5:3:    Field `extends` references unknown identifier
                                    `cwl:ProcessRequirement`, tried
                                    http://commonwl.org/cwltool#cwl:ProcessRequirement
http://commonwl.org/cwltool:6:3:    checking field `fields`
http://commonwl.org/cwltool:22:5:     checking object
                                      `http://commonwl.org/cwltool#CUDARequirement/cudaDeviceCountMax`
                                        Field `type` references unknown identifier
                                        `cwl:Expression`, tried http://commonwl.org/cwltool#cwl:Expression
http://commonwl.org/cwltool:32:5:     checking object
                                      `http://commonwl.org/cwltool#CUDARequirement/cudaDeviceCountMin`
                                        Field `type` references unknown identifier
                                        `cwl:Expression`, tried http://commonwl.org/cwltool#cwl:Expression
```
Instead of the previous (expected) error:
```
schema_salad.exceptions.ValidationException: ../../../../tmp/tmp514w9uk1/test:1:114: checking field `requirements`
../../../../tmp/tmp514w9uk1/test:1:131:   checking item
                                            Field `class` contains undefined reference to
                                            `http://commonwl.org/cwltool#CUDARequirement`
```

Is there something else to adjust on my end, or would that be an issue with the schema definitions (using https://github.com/common-workflow-language/cwltool/blob/3.1.20221201130942/cwltool/extensions-v1.2.yml)? 

Tested CWL: 
```cwl
cwlVersion: v1.2
class: CommandLineTool
baseCommand:
- echo
- test
inputs: []
outputs: []
requirements:
  cwltool:CUDARequirement:
    cudaVersionMin: '10.0'
    cudaComputeCapability: '3.0'
    cudaDeviceCountMin: 1
    cudaDeviceCountMax: 1
$namespaces:
  cwltool: http://commonwl.org/cwltool#
```

Reply:
Perhaps you filtered out too much? https://github.com/common-workflow-language/cwltool/blob/38f42e6c4fe17db5fe82acd8a59f97d3982c605b/cwltool/extensions-v1.2.yml#L1-L5 might be missing due to https://github.com/crim-ca/weaver/pull/510/files#diff-ce41af74805c1d891a393bd1488434a450041ca7cc7df4f3532b4e516badf656R426

Since you are pinning your cwltool version, it might be easier to just make your own copy of https://github.com/common-workflow-language/cwltool/blob/38f42e6c4fe17db5fe82acd8a59f97d3982c605b/cwltool/extensions-v1.2.yml and remove the unsupported extensions

Reply:
The import statement is still in the `$graph`. The only extensions removed are the ones I marked as not yet supported.

![image](https://user-images.githubusercontent.com/19194484/205667993-1fc65811-81c7-4e12-95ce-d6a824774c7f.png)

Reply:
I just found my mistake. I was passing down `$graph` rather than the full definition to `use_custom_schema`. Works as expected now!